lib/awful/layout/init.lua

```diff
@@ -181,14 +181,44 @@ function layout.arrange(screen)
         local p = layout.parameters(nil, screen)
 
         local useless_gap = p.useless_gap
+        local ugap_edge = 0
+        local ugap_d = math.floor(useless_gap/2)
+        local ugap_m = math.fmod(useless_gap,2)
+        local m = 0
+
 
         p.geometries = setmetatable({}, {__mode = "k"})
         layout.get(screen).arrange(p)
         for c, g in pairs(p.geometries) do
-            g.width = math.max(1, g.width - c.border_width * 2 - useless_gap * 2)
-            g.height = math.max(1, g.height - c.border_width * 2 - useless_gap * 2)
-            g.x = g.x + useless_gap
-            g.y = g.y + useless_gap
+            if g.x ~= p.workarea.x
+                then m = ugap_d
+                else m = ugap_edge
+            end
+            g.x = g.x + m
+            g.width = g.width - m
+
+            if (g.x + g.width) ~= (p.workarea.x + p.workarea.width)
+                then m = ugap_d + ugap_m
+                else m = ugap_edge
+            end
+            g.width = g.width - m
+
+            if g.y ~= p.workarea.y 
+                then m = ugap_d
+                else m = ugap_edge
+            end
+            g.y = g.y + m
+            g.height = g.height - m
+
+            if (g.y + g.height) ~= (p.workarea.y + p.workarea.height)
+                then m = ugap_d + ugap_m
+                else m = ugap_edge
+            end
+            g.height = g.height - m
+
+            g.width = g.width - c.border_width * 2 
+            g.height = g.height - c.border_width * 2
```

Comment:
my bad


so i've just merged it locally to test again -- smth got changed (in awesome itself probably) -- now `ugap_edge` is being added to the "outer" gap size

however setting `ugap_edge` to `-useless_gap`



here is the full patch with my proposal fix to make those edges configurable via `beautiful`:

```diff
diff --git a/lib/awful/layout/init.lua b/lib/awful/layout/init.lua
index 5fa5ec3d..e5ddd50f 100644
--- a/lib/awful/layout/init.lua
+++ b/lib/awful/layout/init.lua
@@ -21,6 +21,7 @@ local client = require("awful.client")
 local ascreen = require("awful.screen")
 local timer = require("gears.timer")
 local gmath = require("gears.math")
+local beautiful = require("beautiful")
 
 local function get_screen(s)
     return s and capi.screen[s]
@@ -189,6 +190,7 @@ function layout.arrange(screen)
     delayed_arrange[screen] = true
 
     timer.delayed_call(function()
+        local theme = beautiful.get()
         if not screen.valid then
             -- Screen was removed
             delayed_arrange[screen] = nil
@@ -200,44 +202,56 @@ function layout.arrange(screen)
         local p = layout.parameters(nil, screen)
 
         local useless_gap = p.useless_gap
-        local ugap_edge = 0
+        local ugap_edge_left, ugap_edge_top, ugap_edge_right, ugap_edge_bottom = useless_gap, useless_gap, useless_gap, useless_gap
+        if theme.useless_gap_edge_left ~= nil then
+            ugap_edge_left = theme.useless_gap_edge_left
+        end
+        if theme.useless_gap_edge_top ~= nil then
+            ugap_edge_top = theme.useless_gap_edge_top
+        end
+        if theme.useless_gap_edge_right ~= nil then
+            ugap_edge_right = theme.useless_gap_edge_right
+        end
+        if theme.useless_gap_edge_bottom ~= nil then
+            ugap_edge_bottom = theme.useless_gap_edge_bottom
+        end
+        --local useless_gap = 0
         local ugap_d = math.floor(useless_gap/2)
         local ugap_m = math.fmod(useless_gap,2)
         local m = 0
 
-
         p.geometries = setmetatable({}, {__mode = "k"})
         layout.get(screen).arrange(p)
         for c, g in pairs(p.geometries) do
             if g.x ~= p.workarea.x
                 then m = ugap_d
-                else m = ugap_edge
+                else m = ugap_edge_left - useless_gap
             end
             g.x = g.x + m
             g.width = g.width - m
 
             if (g.x + g.width) ~= (p.workarea.x + p.workarea.width)
                 then m = ugap_d + ugap_m
-                else m = ugap_edge
+                else m = ugap_edge_right - useless_gap
             end
             g.width = g.width - m
 
-            if g.y ~= p.workarea.y 
+            if g.y ~= p.workarea.y
                 then m = ugap_d
-                else m = ugap_edge
+                else m = ugap_edge_top - useless_gap
             end
             g.y = g.y + m
             g.height = g.height - m
 
             if (g.y + g.height) ~= (p.workarea.y + p.workarea.height)
                 then m = ugap_d + ugap_m
-                else m = ugap_edge
+                else m = ugap_edge_bottom - useless_gap
             end
             g.height = g.height - m
 
-            g.width = g.width - c.border_width * 2 
+            g.width = g.width - c.border_width * 2
             g.height = g.height - c.border_width * 2
-            
+
             c:geometry(g)
         end
         arrange_lock = false
```